proposals/overload-resolution-priority.md

```diff
@@ -0,0 +1,214 @@
+>
+> The binding-time processing of a method invocation of the form `M(A)`, where `M` is a method group (possibly including a *type_argument_list*), and `A` is an optional *argument_list*, consists of the following steps:
+> 
+> - The set of candidate methods for the method invocation is constructed. For each method `F` associated with the method group `M`:
+>   - If `F` is non-generic, `F` is a candidate when:
+>     - `M` has no type argument list, and
+>     - `F` is applicable with respect to `A` ([§12.6.4.2](expressions.md#12642-applicable-function-member)).
+>   - If `F` is generic and `M` has no type argument list, `F` is a candidate when:
+>     - Type inference ([§12.6.3](expressions.md#1263-type-inference)) succeeds, inferring a list of type arguments for the call, and
+>     - Once the inferred type arguments are substituted for the corresponding method type parameters, all constructed types in the parameter list of `F` satisfy their constraints ([§8.4.5](types.md#845-satisfying-constraints)), and the parameter list of `F` is applicable with respect to `A` ([§12.6.4.2](expressions.md#12642-applicable-function-member))
+>   - If `F` is generic and `M` includes a type argument list, `F` is a candidate when:
+>     - `F` has the same number of method type parameters as were supplied in the type argument list, and
+>     - Once the type arguments are substituted for the corresponding method type parameters, all constructed types in the parameter list of `F` satisfy their constraints ([§8.4.5](types.md#845-satisfying-constraints)), and the parameter list of `F` is applicable with respect to `A` ([§12.6.4.2](expressions.md#12642-applicable-function-member)).
+> - The set of candidate methods is reduced to contain only methods from the most derived types: For each method `C.F` in the set, where `C` is the type in which the method `F` is declared, all methods declared in a base type of `C` are removed from the set. Furthermore, if `C` is a class type other than `object`, all methods declared in an interface type are removed from the set.  
+>   > *Note*: This latter rule only has an effect when the method group was the result of a member lookup on a type parameter having an effective base class other than `object` and a non-empty effective interface set. *end note*
+> - If the resulting set of candidate methods is empty, then further processing along the following steps are abandoned, and instead an attempt is made to process the invocation as an extension method invocation ([§12.8.9.3](expressions.md#12893-extension-method-invocations)). If this fails, then no applicable methods exist, and a binding-time error occurs.
+> - **The resulting set of candidate methods is grouped by ***overload_resolution_priority***. All candidates not in the highest (by standard integer comparison) ***overload_resolution_priority*** group are removed from the set.**
```

Comment:
First, I think that a better place for the change is actually in Overload Resolution part of the language spec. In [12.6.4.1 General](https://github.com/dotnet/csharpstandard/blob/draft-v9/standard/expressions.md#12641-general). Perhaps a new bullet point right after:
>- First, the set of candidate function members is reduced to those function members that are applicable with respect to the given argument list ([§12.6.4.2](https://github.com/dotnet/csharpstandard/blob/draft-v9/standard/expressions.md#12642-applicable-function-member)). If this reduced set is empty, a compile-time error occurs.



Second, I think the proposed wording of the bullet point doesn't accurately reflect the intent of filtering per declaring type: "adjust the relative priority of overloads within a single type". 


It should probably say some thing like: "Group candidates per declaring type. Determine the highest overload_resolution_priority in each group. Candidates with lower overload_resolution_priority from each group are removed from the set of candidates."